### ISSUE TITLE
`ClampingScrollSimulation` has large nonzero velocity when near the end, and velocity even increases slightly there

### ISSUE THREAD

**Issue by fzyzcjy** [flutter/flutter#113424]
## The problem

When using ClampingScrollSimulation, there are two problems:

* It has large nonzero velocity when near the end (see figure below). We all know, for list view scrolling to feel comfortable, it should decrease its velocity and get to near-zero value when near end, instead of a large nonzero value.
* velocity even increases slightly there (again see figure below). We know velocity should at least decrease due to friction.

## Reproduction code for "velocity increases when near end"

```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('ClampingScrollSimulation should have decreasing speed', () {
    final ClampingScrollSimulation simulation =
        ClampingScrollSimulation(position: 0, velocity: 1000);

    final List<double> tValues = List<double>.generate(60, (int i) => i / 60);
    final List<double> xValues = tValues.map(simulation.x).toList();
    final List<double> xDeltas = <double>[
      for (int i = 0; i < xValues.length - 1; ++i) xValues[i + 1] - xValues[i]
    ];

    // If curious, uncomment this to see the real values
    // debugPrint('$tValues\n$xValues\n$xDeltas');

    for (int i = 0; i < xDeltas.length - 1; ++i) {
      expect(xDeltas[i], greaterThan(xDeltas[i + 1]));
    }
  });
}
```

## Figures

Code:

<details>

```python
import matplotlib
import numpy as np

matplotlib.use("MacOSX")

import matplotlib.pyplot as plt

import math
from dataclasses import dataclass

_kDecelerationRate = math.log(0.78) / math.log(0.9)
_initialVelocityPenetration = 3.065


def _decelerationForFriction(friction: float):
    return friction * 61774.04968


def _flingDistancePenetration(t):
    return (1.2 * t * t * t) - (3.27 * t * t) + (_initialVelocityPenetration * t)


def _flingVelocityPenetration(t):
    return (3.6 * t * t) - (6.54 * t) + _initialVelocityPenetration


def clampDouble(x, min, max):
    if x < min:
        return min
    if x > max:
        return max
    return x


def _sign(x):
    if x > 0:
        return 1.0
    if x < 0:
        return -1.0
    return x


def _array_map(x, f):
    return np.fromiter((f(xi) for xi in x), x.dtype)


@dataclass
class ClampingScrollSimulation:
    position: float
    velocity: float
    friction: float = 0.015

    def __post_init__(self):
        self._duration = self._flingDuration(self.velocity)
        self._distance = abs(self.velocity * self._duration / _initialVelocityPenetration)

    def _flingDuration(self, velocity: float):
        scaledFriction = self.friction * _decelerationForFriction(0.84)
        deceleration = math.log(0.35 * abs(velocity) / scaledFriction)
        return math.exp(deceleration / (_kDecelerationRate - 1.0))

    def x(self, time):
        t = clampDouble(time / self._duration, 0.0, 1.0)
        return self.position + self._distance * _flingDistancePenetration(t) * _sign(self.velocity)

    def dx(self, time):
        t = clampDouble(time / self._duration, 0.0, 1.0)
        return self._distance * _flingVelocityPenetration(t) * _sign(self.velocity) / self._duration


# %%

plt.clf()
plt.tight_layout()
ax1 = plt.gca()
ax2 = ax1.twinx()

dt = 1 / 60
t = np.arange(0, 1, dt)

simulation_a = ClampingScrollSimulation(position=0, velocity=1000)
x_a = _array_map(t, simulation_a.x)
v_a = _array_map(t, simulation_a.dx)
diffx_a = (x_a[1:] - x_a[:-1]) / dt

ax1.plot(t, x_a, label='x_a')
ax2.plot(t, v_a, label='v_a')
ax2.plot(t[:-1], diffx_a, label='diffx_a')

if False:
    b_start_time = 0.3
    simulation_b = ClampingScrollSimulation(
        position=simulation_a.x(b_start_time),
        velocity=simulation_a.dx(b_start_time),
    )
    x_b = _array_map(t, simulation_b.x)
    v_b = _array_map(t, simulation_b.dx)
    diffx_b = (x_b[1:] - x_b[:-1]) / dt

    ax1.plot(t + b_start_time, x_b, label='x_b')
    ax2.plot(t + b_start_time, v_b, label='v_b')
    ax2.plot(t[:-1] + b_start_time, diffx_b, label='diffx_b')

ax1.legend(loc="upper left")
ax2.legend(loc="upper right")

plt.xlim([0, 1])

plt.show()

```

</details>

Image:

![image](https://user-images.githubusercontent.com/5236035/195789813-b1131a0a-ed52-4111-9bfb-030ae5133fc6.png)

zoom in:

![image](https://user-images.githubusercontent.com/5236035/195789952-6dcaad63-0759-4da4-8441-74a85944e5d6.png)

zoom in again:

![image](https://user-images.githubusercontent.com/5236035/195789996-7aa7eb3d-cb94-4aa1-981c-23a8b3c47ec7.png)

**Comment by huycozy** [comment]
Hi @fzyzcjy
I label this issue for further investigation/insight from the team. This issue is reproducible on the latest `stable` and `master` channels with provided sample code.

<details>
<summary>Test result output</summary>

```bash
Expected: a value greater than <0.5439357546201506>
  Actual: <0.51773409104743>
   Which: is not a value greater than <0.5439357546201506>
```
</details>

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.3.4, on macOS 12.6 21G115 darwin-x64, locale en-VN)
    • Flutter version 3.3.4 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision eb6d86ee27 (23 hours ago), 2022-10-04 22:31:45 -0700
    • Engine revision c08d7d5efc
    • Dart version 2.18.2
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode14.0.1.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.71.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.48.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.6 21G115 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 105.0.5195.125

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```

```bash
[✓] Flutter (Channel master, 3.5.0-6.0.pre.37, on macOS 12.6 21G115 darwin-x64, locale en-VN)
    • Flutter version 3.5.0-6.0.pre.37 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 08e1729204 (9 hours ago), 2022-10-13 17:40:07 -0400
    • Engine revision 3711bbaeab
    • Dart version 2.19.0 (build 2.19.0-309.0.dev)
    • DevTools version 2.18.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode14.0.1.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.72.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • iPhone 14 Pro Max (mobile) • 332079AF-E895-4685-910E-7B1E18B0C6B8 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)            • macos                                • darwin-x64     • macOS 12.6 21G115 darwin-x64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 106.0.5249.119

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```
</details>

**Comment by Piinks** [comment]
@fzyzcjy I am not sure what these charts are showing, can you explain?

**Comment by Piinks** [comment]
Is this a duplicate of https://github.com/flutter/flutter/issues/16371?

**Comment by fzyzcjy** [comment]
@Piinks The charts are reproduced from the attached python code. The python code is just a python version of dart ClampingScrollSimulation (because I want to use python's plotting lib).

That may not be a duplicate. Because that issue is talking about "not matching" (which is quite strict), while this issue is merely saying it is counter-intuitive.

**Comment by Piinks** [comment]
> The charts are reproduced from the attached python code. The python code is just a python version of dart ClampingScrollSimulation (because I want to use python's plotting lib).

Thanks for the context! Can you add labels? I don't know or understand what each line or axis represents.

**Comment by gnprice** [comment]
Here's a repro recipe for this issue, directly in terms of how a user sees it:

## Steps to Reproduce

1. Run the [`platform_tests/scroll_overlay`](https://github.com/flutter/platform_tests/tree/1d797b8699546f1178e3cf58de9c6cc64f825281/scroll_overlay) app on an Android device. (Or really any other app with a long, scrollable list using the default behavior, on any non-Apple platform.)
2. Swipe up rapidly, once, to scroll with a rapid fling. Wait a couple of seconds as the scrolling slows down.
3. Watch carefully in the last quarter-second or so before the scrolling completes.

**Expected results:** 

The scrolling motion should continuously slow down until it stops.

**Actual results:** 

After slowing down for a while, the scrolling motion speeds up again slightly, and then abruptly stops.

**Comment by gnprice** [comment]
This issue has been reported several times in passing, as part of other issues or PRs. I think this is the right thread to use to track it (and to consolidate any future duplicates onto.)

Recording those here for reference:

* #77497
  > I've noticed that the Android Flutter scroll logic appears to be slightly off. After a scroll slows down to a sufficient amount, items appear to "snap" into place. It's as if the inertial scrolling clamps down and micro movements are not allowed. Note the difference is very subtle but present across Flutter apps, e.g. Stadia, gPay etc. […] Since this has been driving me mad […]

* PR aimed at specifically this issue, with a repro video: #92338

* One of several observations in: #16371

  > Besides, before the end of scroll, I have a feeling that it "suddenly" stops.

* And last week by me at https://github.com/flutter/flutter/issues/119875#issuecomment-1414580430 .

Copy-pasting from that last one the part that belongs on this issue instead, and editing slightly:

An issue with the `ClampingScrollSimulation` curve: the scroll does not come to a smooth stop.

Instead the velocity actually increases over the last 200ms or so, before abruptly stopping when the simulation ends. In the Android scroll physics, the velocity decreases continuously to reach zero at the end.

I originally noticed this issue thanks to the graphs I posted at https://github.com/flutter/flutter/issues/119875#issuecomment-1416241372 ; when interacting as a user, I hadn't ever put my finger on it. But now that I know about it, it's very conspicuous to me when using `scroll_overlay` and I find it noticeable even in a normal app with a long list to scroll through.

I suspect that for lots of users the lack of a smooth stop contributes to a feeling that the scrolling isn't quite right, even without them ever identifying specifically why.

To fix this issue and get a smooth stop, we want the curve to have velocity 0 at t = 1, i.e. dx/dt(1) = 0, like the Android curve does.

**Comment by gnprice** [comment]
Rereading this now, I have a correction to make to my previous comment:

> 1. Run the [`platform_tests/scroll_overlay`](https://github.com/flutter/platform_tests/tree/1d797b8699546f1178e3cf58de9c6cc64f825281/scroll_overlay) app on an Android device. (Or really any other app with a long, scrollable list using the default behavior, on any non-Apple platform.)

This issue did not actually reproduce on `platform_tests/scroll_overlay` out of the box. That's because out of the box (and before #120420 was merged just now), that app showed the symptoms of #120338: the scroll fling simulation would be frequently restarted, which masked this issue.

I'd been seeing this on a version of that app where I'd made the list items all have the same height, instead of each item being slightly taller than the last. That caused it not to trigger #120338, so that it did show the symptoms of this issue. I guess when I wrote the above comment, I hadn't yet fully understood how these two issues interacted.

In any case, #120420 should fix both of these issues and has now been merged.
